theano/tensor/nnet/nnet.py

```diff
@@ -343,68 +331,80 @@ def c_code(self, node, name, inp, out, sub):
-            npy_intp Sdy = PyArray_STRIDES(%(dy)s)[1]/sizeof(dtype_%(dy)s);
-            const dtype_%(sm)s* __restrict__ sm_i = (dtype_%(sm)s*) (PyArray_BYTES(%(sm)s) + PyArray_STRIDES(%(sm)s)[0] * i);
-            npy_intp Ssm = PyArray_STRIDES(%(sm)s)[1]/sizeof(dtype_%(sm)s);
-            dtype_%(dx) s* __restrict__ dx_i = (dtype_%(dx)s*) (PyArray_BYTES(%(dx)s) + PyArray_STRIDES(%(dx)s)[0] * i);
-            npy_intp Sdx = PyArray_STRIDES(%(dx)s)[1]/sizeof(dtype_%(dx)s);
-
-            double sum_dy_times_sm = 0.;
-            for (size_t j = 0; j < PyArray_DIMS(%(dx)s)[1]; ++j)
+            npy_intp* input_shape = PyArray_DIMS(%(sm)s);
+            npy_int ndim = PyArray_NDIM(%(sm)s);
+            if ((PyArray_TYPE(%(dy)s) != NPY_DOUBLE) &&
+                    (PyArray_TYPE(%(dy)s) != NPY_FLOAT))
+                {
+                    PyErr_SetString(PyExc_TypeError,
+                        "types should be float or float64");
+                    %(fail)s;
+                }
+
+            // Check input type
+            if ((PyArray_TYPE(%(sm)s) != NPY_DOUBLE) &&
+                    (PyArray_TYPE(%(sm)s) != NPY_FLOAT))
             {
-                dx_i[j * Sdx] = dy_i[j * Sdy] * sm_i[j * Ssm];
-                sum_dy_times_sm += dx_i[j * Sdx];
-            }
-            for (size_t j = 0; j < PyArray_DIMS(%(dx)s)[1]; ++j)
+                    PyErr_SetString(PyExc_TypeError,
+                        "Types should be float or float64");
+                    %(fail)s;
+                    }
+
+            // Check the number of dimension
+            if (PyArray_NDIM(%(dy)s) != PyArray_NDIM(%(sm)s))
+                {
+                    PyErr_SetString(PyExc_ValueError, "Sm and dy should have the same number of dimension.");
+                    %(fail)s;
+                }
+
+            // Check the shapes
+            if (PyArray_DIMS(%(dy)s) == PyArray_DIMS(%(sm)s))
+                {
+                    PyErr_SetString(PyExc_ValueError, "Sm and dy should have the same shape.");
+                    %(fail)s;
+                }
+
+            // Alloc memory for the output
+            if ((NULL == %(dx)s) || (PyArray_DIMS(%(dx)s) != PyArray_DIMS(%(sm)s)))
+                {
+                    Py_XDECREF(%(dx)s);
+                    %(dx)s = (PyArrayObject*) PyArray_SimpleNew(PyArray_NDIM(%(sm)s),
+                                                                PyArray_DIMS(%(sm)s),
+                                                                PyArray_TYPE(%(sm)s));
+                    if (!%(dx)s)
+                    {
+                        PyErr_SetString(PyExc_MemoryError,
+                            "failed to alloc dx output");
+                        %(fail)s;
+                    }
+                }
+
+            // Compute array size except on the last dimension
+            npy_int input_size;
+            input_size = PyArray_SIZE(%(sm)s) / input_shape[ndim - 1];
```

Comment:
This could crash the program if `input_shape[ndim - 1]` is equal to 0.
Could you detect that case and raise a Python exception instead? It makes things much easier to debug.